File: Services/pkg/user_role/cache.go
package user_role

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/ChangerzaryX1602/SkillSync/pkg/models"
	"github.com/ChangerzaryX1602/SkillSync/pkg/utils"
	"github.com/gofiber/storage/redis"
)

const (
	cacheTTL     = 15 * time.Minute
	cacheTTLList = 1 * time.Minute
)

type userRoleCache struct {
	store *redis.Storage
}

func newUserRoleCache(store *redis.Storage) *userRoleCache {
	return &userRoleCache{store: store}
}

func (c *userRoleCache) GetByUserID(userID uint) ([]models.UserRole, bool) {
	if c.store == nil {
		return nil, false
	}
	key := fmt.Sprintf("%s:%d", models.PkgUserRoleGetByUserID, userID)

	bytes, err := c.store.Get(key)
	if err != nil || len(bytes) == 0 {
		return nil, false
	}

	var userRoles []models.UserRole
	if err := json.Unmarshal(bytes, &userRoles); err != nil {
		return nil, false
	}
	return userRoles, true
}

func (c *userRoleCache) SetByUserID(userID uint, userRoles []models.UserRole) {
	if c.store == nil {
		return
	}
	go func(urs []models.UserRole) {
		key := fmt.Sprintf("%s:%d", models.PkgUserRoleGetByUserID, userID)
		if bytes, err := json.Marshal(urs); err == nil {
			_ = c.store.Set(key, bytes, utils.RandomJitter(cacheTTLList))
		}
	}(userRoles)
}

func (c *userRoleCache) InvalidateByUserID(userID uint) {
	if c.store == nil {
		return
	}
	go func() {
		key := fmt.Sprintf("%s:%d", models.PkgUserRoleGetByUserID, userID)
		_ = c.store.Delete(key)
	}()
}

File: hono-clean-architecture/pkg/user_role/cache.ts
import Redis from "ioredis";
import { UserRole } from "../models";
import { randomJitter } from "../utils";

const CACHE_TTL = 15 * 60 * 1000;

export class UserRoleCache {
  private store: Redis | null;

  constructor(store: Redis | null) {
    this.store = store;
  }

  async get(id: number): Promise<UserRole | null> {
    if (!this.store) return null;

    const key = `pkg:userrole:get:${id}`;
    const data = await this.store.get(key);

    if (!data) return null;

    try {
      return JSON.parse(data) as UserRole;
    } catch {
      return null;
    }
  }

  async set(userRole: UserRole): Promise<void> {
    if (!this.store || !userRole) return;

    const key = `pkg:userrole:get:${userRole.id}`;
    const ttl = randomJitter(CACHE_TTL);

    await this.store.set(key, JSON.stringify(userRole), "PX", ttl);
  }

  async invalidate(id: number): Promise<void> {
    if (!this.store) return;
    await this.store.del(`pkg:userrole:get:${id}`);
  }

  async invalidateAllLists(): Promise<void> {
    if (!this.store) return;

    const pattern = `pkg:userrole:list:*`;
    const keys = await this.store.keys(pattern);

    if (keys.length > 0) {
      await this.store.del(...keys);
    }
  }
}

export function newUserRoleCache(store: Redis | null): UserRoleCache {
  return new UserRoleCache(store);
}
